pyproject.toml

```diff
@@ -30,9 +31,129 @@ classifiers = [
     "Operating System :: MacOS",
     "Programming Language :: Python :: 3",
 ]
+scripts = { mne = "mne.commands.utils:main" }
+dependencies = [
+    "numpy>=1.21.2",
+    "scipy>=1.7.1",
+    "matplotlib>=3.5.0",
+    "tqdm",
+    "pooch>=1.5",
+    "decorator",
+    "packaging",
+    "jinja2",
+    "importlib_resources>=5.10.2; python_version<'3.9'",
+    "lazy_loader>=0.3",
+    "defusedxml",
+]
+
+[project.optional-dependencies]
+# Variants with dependencies that will get installed on top of those listed unter
+# project.dependencies
+
+# Leave this one here for backward-compat
+data = []
+
+# Dependencies for MNE-Python functions that use HDF5 I/O
+hdf5 = ["h5io", "pymatreader"]
+
+# Dependencies for full MNE-Python functionality (other than raw/epochs export)
+full = [
+    "mne[hdf5]",
+    "qtpy",
+    "PyQt6",
+    "pyobjc-framework-Cocoa>=5.2.0; platform_system=='Darwin'",
+    "sip",
+    "scikit-learn",
+    "nibabel",
+    "openmeeg>=2.5.5",
+    "numba",
+    "h5py",
+    "pandas",
+    "numexpr",
+    "jupyter",
+    "python-picard",
+    "statsmodels",
+    "joblib",
+    "psutil",
+    "dipy",
+    "vtk",
+    "nilearn",
+    "xlrd",
+    "imageio>=2.6.1",
+    "imageio-ffmpeg>=0.4.1",
+    "traitlets",
+    "pyvista>=0.32,!=0.35.2,!=0.38.0,!=0.38.1,!=0.38.2,!=0.38.3,!=0.38.4,!=0.38.5,!=0.38.6,!=0.42.0",
+    "pyvistaqt>=0.4",
+    "mffpy>=0.5.7",
+    "ipywidgets",
+    "ipympl",
+    "ipyevents",
+    "trame",
+    "trame-vtk",
+    "trame-vuetify",
+    "mne-qt-browser",
+    "darkdetect",
+    "qdarkstyle",
+    "threadpoolctl",
+]
+
+# Dependencies for running the test infrastructure
+test = [
+    "pytest",
+    "pytest-cov",
+    "pytest-timeout",
+    "pytest-harvest",
+    "pytest-qt",
+    "ruff",
+    "numpydoc",
+    "codespell",
+    "check-manifest",
+    "tomli; python_version<'3.11'",
+    "twine",
+    "wheel",
+    "pre-commit",
+    "black",
+]
 
-[project.scripts]
-mne = "mne.commands.utils:main"
+# Dependencies for being able to run all tests (rare/CIs/advanced devs)
+test_full = [
+    "mne[full]",
```

Comment:
Reorganizing this way breaks the `--pre` CI build, because `full` pulls in `pandas` which pins NumPy<2. So I think we should go back to having `test_extra` rather than `test_full`. Then in CI jobs use `[full,test_extra]` to mean what `[test_full]` means now